analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/HashesShouldHaveUnpredictableSaltBase.cs

```diff
@@ -18,12 +18,89 @@
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        var instance = context.Operation.Instance;
+        if (instance.AsObjectCreation() is { } objectCreation)
+        {
+            return ProcessObjectCreation(objectCreation, state);
+        }
+        else if (instance.AsInvocation() is { } invocation)
+        {
+            return ProcessInvocation(invocation, state);
+        }
+        else if (instance.AsArrayCreation() is { } arrayCreation)
+        {
+            return ProcessArrayCreation(arrayCreation, state);
+        }
+        else
+        {
+            return state;
+        }
+    }
+
+    private ProgramState ProcessObjectCreation(IObjectCreationOperationWrapper objectCreation, ProgramState state)
+    {
+        if (objectCreation.Type.DerivesFrom(KnownType.System_Security_Cryptography_DeriveBytes)
+            && objectCreation.Arguments.FirstOrDefault(x => x.AsArgument() is { Parameter.Name: "salt" or "rgbSalt" }) is { } saltArgument)
+        {
+            if (state[saltArgument]?.HasConstraint(ByteCollectionConstraint.CryptographicallyWeak) is true)
+            {
+                ReportIssue(saltArgument, MakeSaltUnpredictableMessage);
+            }
+            else if (state[saltArgument]?.HasConstraint(SaltSizeConstraint.Short) is true)
+            {
+                ReportIssue(saltArgument, MakeThisSaltLongerMessage);
+            }
+        }
+        return state;
+    }
+
+    private static ProgramState ProcessInvocation(IInvocationOperationWrapper invocation, ProgramState state)
+    {
+        if (CryptographicallyStrongRandomNumberGenerators.Any(x => IsInvocationToRandomNumberGenerator(x, invocation))
+            && invocation.ArgumentValue("data") is { } dataArgument
+            && dataArgument.TrackedSymbol() is { } trackedSymbol)
+        {
+            state = state.SetSymbolConstraint(trackedSymbol, ByteCollectionConstraint.CryptographicallyStrong);
+        }
+        return state;
+    }
+
+    private static ProgramState ProcessArrayCreation(IArrayCreationOperationWrapper arrayCreation, ProgramState state)
+    {
+        if (arrayCreation.Type.Is(KnownType.System_Byte_Array) && arrayCreation.DimensionSizes.Length == 1)
+        {
+            state = state.SetOperationConstraint(arrayCreation.WrappedOperation, ByteCollectionConstraint.CryptographicallyWeak);
+
+            if (arrayCreation.DimensionSizes.Single().ConstantValue.Value is int arraySize && arraySize < SafeSaltSize)
+            {
+                state = state.SetOperationConstraint(arrayCreation.WrappedOperation, SaltSizeConstraint.Short);
+            }
+        }
+        return state;
+    }
+
+    private static bool IsInvocationToRandomNumberGenerator(MemberDescriptor methodDescriptor, IInvocationOperationWrapper invocation) =>
```

Comment:
this can be deleted with the change I proposed earlier.

Reply:
See the response for the previous comment.